jobs/picard.py

```diff
@@ -65,3 +65,178 @@ def markdup(
             )
 
     return j
+
+
+def picard_collect_metrics(
+    b,
+    cram_path: CramPath,
+    out_alignment_summary_metrics_path: Path,
+    out_base_distribution_by_cycle_metrics_path: Path,
+    out_insert_size_metrics_path: Path,
+    out_quality_by_cycle_metrics_path: Path,
+    out_quality_yield_metrics_path: Path,
+    job_attrs: dict | None = None,
+    overwrite: bool = False,
+) -> Job | None:
+    """
+    Run picard CollectMultipleMetrics metrics for sample QC.
+    Based on https://github.com/broadinstitute/warp/blob/master/tasks/broad/Qc.wdl#L141
+    """
+    if can_reuse(
+        [
+            out_alignment_summary_metrics_path,
+            out_base_distribution_by_cycle_metrics_path,
+            out_insert_size_metrics_path,
+            out_quality_by_cycle_metrics_path,
+            out_quality_yield_metrics_path,
+        ],
+        overwrite,
+    ):
+        return None
+
+    job_attrs = (job_attrs or {}) | {'tool': 'picard_CollectMultipleMetrics'}
+    j = b.new_job('Picard CollectMultipleMetrics', job_attrs)
+    j.image(image_path('picard'))
+    res = STANDARD.request_resources(mem_gb=7)
+    res.attach_disk_storage_gb = storage_for_cram_qc_job()
+    res.set_to_job(j)
+    cram = cram_path.resource_group(b)
+    reference = fasta_res_group(b)
+
+    cmd = f"""\
+    picard -Xms5g -Xmx{res.get_java_mem_mb()}m \
+      CollectMultipleMetrics \
+      INPUT={cram.cram} \
+      REFERENCE_SEQUENCE={reference.base} \
+      OUTPUT=$BATCH_TMPDIR/prefix \
+      ASSUME_SORTED=true \
+      PROGRAM=null \
+      PROGRAM=CollectAlignmentSummaryMetrics \
+      PROGRAM=CollectInsertSizeMetrics \
+      PROGRAM=MeanQualityByCycle \
+      PROGRAM=CollectBaseDistributionByCycle \
+      PROGRAM=CollectQualityYieldMetrics \
+      METRIC_ACCUMULATION_LEVEL=null \
+      METRIC_ACCUMULATION_LEVEL=SAMPLE
+      
+    ls $BATCH_TMPDIR/
+    cp $BATCH_TMPDIR/prefix.alignment_summary_metrics {j.out_alignment_summary_metrics}
+    cp $BATCH_TMPDIR/prefix.base_distribution_by_cycle_metrics {j.out_base_distribution_by_cycle_metrics}
+    cp $BATCH_TMPDIR/prefix.insert_size_metrics {j.out_insert_size_metrics}
+    cp $BATCH_TMPDIR/prefix.quality_by_cycle_metrics {j.out_quality_by_cycle_metrics}
+    cp $BATCH_TMPDIR/prefix.quality_yield_metrics {j.out_quality_yield_metrics}
+    """
+
+    j.command(command(cmd))
+    b.write_output(
+        j.out_alignment_summary_metrics, str(out_alignment_summary_metrics_path)
+    )
+    b.write_output(j.out_insert_size_metrics, str(out_insert_size_metrics_path))
+    b.write_output(
+        j.out_quality_by_cycle_metrics, str(out_quality_by_cycle_metrics_path)
+    )
+    b.write_output(
+        j.out_base_distribution_by_cycle_metrics,
+        str(out_base_distribution_by_cycle_metrics_path),
+    )
+    b.write_output(j.out_quality_yield_metrics, str(out_quality_yield_metrics_path))
```

Comment:
This seems less clean than creating a single resource group for the outputs, but I guess it's necessary if all the outputs need to end up on different paths

Reply:
e.g. https://github.com/populationgenomics/rare-disease/blob/main/validation/validation_runner.py#L171-L194

Reply:
Yeah, I thought about changing it to a resource group, but I wanted flexibility of writing to different paths, plus I find this a bit more clear personally because I always forget how resource groups actually work :)

Reply:
Same 😆